File: php/tests/BarcodeTransferTest.php
<?php

declare(strict_types=1);

namespace Ingot\Tests;

use Ingot\Claim;
use Ingot\Cluster;
use Ingot\ConflictFlagged;
use Ingot\Events;
use Ingot\IdentityLedger;
use Ingot\LedgerState;
use Ingot\Substrate;
use PHPUnit\Framework\TestCase;

/**
 * PHP parity for test/barcode_transfer_test.exs.
 *
 * A barcode is reassignable — NHSBSA publishes a GTIN Transfer Tracking Log of codes moving
 * between dm+d packs — so a surrogate key must never follow one. National codes, assigned once
 * and never reissued, decide which cluster continues the key.
 */
final class BarcodeTransferTest extends TestCase
{
    private const GTIN = ['gtin', '05012345678900'];

    public function test_the_key_stays_with_the_pack_that_keeps_its_national_code(): void
    {
        $ledger = $this->week(IdentityLedger::new(), [
            $this->listing('100g', [['cnk', '1111111'], self::GTIN]),
        ]);

        self::assertSame(['SK_1'], array_keys($ledger->members));

        $ledger = $this->week($ledger, [
            $this->listing('100g', [['cnk', '1111111']]),
            $this->listing('30g', [['cnk', '2222222'], self::GTIN]),
        ]);

        self::assertSame([['cnk', '1111111']], $this->codes($ledger, 'SK_1'));
        self::assertSame(
            [['cnk', '2222222'], ['gtin', '05012345678900']],
            $this->codes($ledger, 'SK_2')
        );
    }

    public function test_with_no_national_codes_the_key_stays_with_its_own_reference(): void
    {
        $ledger = $this->week(IdentityLedger::new(), [
            $this->listing('100g', [self::GTIN, ['supplier_ref', 'AMPP-100g']]),
        ]);

        $ledger = $this->week($ledger, [
            $this->listing('100g', [['supplier_ref', 'AMPP-100g']]),
            $this->listing('30g', [self::GTIN, ['supplier_ref', 'AMPP-30g']]),
        ]);

        self::assertSame([['supplier_ref', 'AMPP-100g']], $this->codes($ledger, 'SK_1'));
    }

    public function test_the_outcome_does_not_depend_on_listing_order(): void
    {
        $before = $this->week(IdentityLedger::new(), [
            $this->listing('100g', [['cnk', '1111111'], self::GTIN]),
        ]);

        $after = [
            $this->listing('100g', [['cnk', '1111111']]),
            $this->listing('30g', [['cnk', '2222222'], self::GTIN]),
        ];

        $forward = $this->week($before, $after);
        $reversed = $this->week($before, array_reverse($after));

        self::assertSame([['cnk', '1111111']], $this->codes($forward, 'SK_1'));
        self::assertSame($this->allCodes($forward), $this->allCodes($reversed));
    }

    public function test_replacing_the_national_code_on_a_live_key_is_flagged(): void
    {
        $ledger = $this->week(IdentityLedger::new(), [
            $this->listing('r', [['cnk', '1111111'], self::GTIN]),
        ]);

        $events = $this->decide($ledger, [$this->listing('r', [['cnk', '2222222'], self::GTIN])]);

        self::assertTrue($this->hasSwapFlag($events), 'expected an identity_swap flag');
    }

    public function test_gaining_a_second_national_code_is_not_flagged(): void
    {
        $ledger = $this->week(IdentityLedger::new(), [
            $this->listing('r', [['cnk', '1111111'], self::GTIN]),
        ]);

        $events = $this->decide($ledger, [
            $this->listing('r', [['cnk', '1111111'], ['cnk', '2222222'], self::GTIN]),
        ]);

        self::assertFalse($this->hasSwapFlag($events));
    }

    // ── helpers ──────────────────────────────────────────────────────────────

    /** @param list<array{0: string, 1: string}> $codes */
    private function listing(string $ref, array $codes): Claim
    {
        return Substrate::claim('nhsbsa', 'identity', ['ref' => $ref, 'codes' => $codes], 'd1', 'd1');
    }

    /**
     * @param list<Claim> $claims
     *
     * @return list<\Ingot\DomainEvent>
     */
    private function decide(LedgerState $ledger, array $claims): array
    {
        $stamped = [];
        $i = 1;
        foreach ($claims as $claim) {
            $stamped[] = $claim->withOrder($i++);
        }

        return IdentityLedger::decide(
            $ledger,
            ['reconcile', Cluster::variants(Substrate::current($stamped)), 'd1']
        );
    }

    /** @param list<array<string,mixed>> $claims */
    private function week(LedgerState $ledger, array $claims): LedgerState
    {
        foreach ($this->decide($ledger, $claims) as $event) {
            $ledger = IdentityLedger::evolve($ledger, $event);
        }

        return $ledger;
    }

    /** @return list<array{0: string, 1: string}> */
    private function codes(LedgerState $ledger, string $key): array
    {
        $codes = array_values($ledger->members[$key] ?? []);
        sort($codes);

        return $codes;
    }

    /** @return array<string, list<array{0: string, 1: string}>> */
    private function allCodes(LedgerState $ledger): array
    {
        $out = [];
        foreach (array_keys($ledger->members) as $key) {
            $out[$key] = $this->codes($ledger, $key);
        }
        ksort($out);

        return $out;
    }

    /** @param list<\Ingot\DomainEvent> $events */
    private function hasSwapFlag(array $events): bool
    {
        foreach ($events as $event) {
            if ($event instanceof ConflictFlagged && ($event->subject[0] ?? null) === 'identity_swap') {
                return true;
            }
        }

        return false;
    }
}

File: test/barcode_transfer_test.exs
# test/barcode_transfer_test.exs
#
# A barcode is not a permanent name for a product. NHSBSA publishes a GTIN Transfer Tracking Log
# recording GTINs that moved from one dm+d pack to another — because the pack changed size, because
# the code was recorded against the wrong product, or because a company was taken over.
#
# So a surrogate key must never follow a barcode. It belongs to whichever cluster CONTINUES what
# the key already meant, and national codes (assigned once, never reissued) decide that — not the
# barcode, which is explicitly reassignable.

alias GoldenRecord.{Events, Codes, Substrate, Cluster, IdentityLedger}

defmodule BarcodeTransferTest do
  use ExUnit.Case, async: true
  import Substrate, only: [claim: 5]

  @w1 ~D[2026-01-05]
  @w2 ~D[2026-01-12]
  @gtin {:gtin, "05012345678900"}

  describe "a transferred barcode does not take the surrogate key with it" do
    test "the key stays with the pack that keeps its national code" do
      # Week 1: SK_1 IS the 100 g pack — cnk 1111111, carrying the barcode.
      ledger = week(IdentityLedger.new(), [listing("100g", [{:cnk, "1111111"}, @gtin])], @w1)
      assert ledger.members == %{"SK_1" => codes([{:cnk, "1111111"}, @gtin])}

      # Week 2: the barcode moves to the 30 g pack, which has its own national code.
      ledger =
        week(
          ledger,
          [
            listing("100g", [{:cnk, "1111111"}]),
            listing("30g", [{:cnk, "2222222"}, @gtin])
          ],
          @w2
        )

      # SK_1 must still mean the 100 g pack. The barcode left; the key did not go with it.
      assert ledger.members["SK_1"] == codes([{:cnk, "1111111"}])
      assert ledger.members["SK_2"] == codes([{:cnk, "2222222"}, @gtin])
    end

    test "with no national codes anywhere, the key stays with the pack keeping its own reference" do
      ledger =
        week(
          IdentityLedger.new(),
          [listing("100g", [@gtin, {:supplier_ref, "AMPP-100g"}])],
          @w1
        )

      assert ledger.members == %{"SK_1" => codes([@gtin, {:supplier_ref, "AMPP-100g"}])}

      ledger =
        week(
          ledger,
          [
            listing("100g", [{:supplier_ref, "AMPP-100g"}]),
            listing("30g", [@gtin, {:supplier_ref, "AMPP-30g"}])
          ],
          @w2
        )

      # supplier_ref is not a reassignable barcode, so it outranks the GTIN for continuity.
      assert ledger.members["SK_1"] == codes([{:supplier_ref, "AMPP-100g"}])
    end

    test "the outcome does not depend on the order the listings arrive in" do
      before = week(IdentityLedger.new(), [listing("100g", [{:cnk, "1111111"}, @gtin])], @w1)

      after_transfer = [
        listing("100g", [{:cnk, "1111111"}]),
        listing("30g", [{:cnk, "2222222"}, @gtin])
      ]

      forward = week(before, after_transfer, @w2)
      reversed = week(before, Enum.reverse(after_transfer), @w2)

      assert forward.members == reversed.members
      assert forward.members["SK_1"] == codes([{:cnk, "1111111"}])
    end
  end

  describe "an established key that loses a national code is not allowed to change quietly" do
    test "replacing the national code on a live key is flagged" do
      ledger = week(IdentityLedger.new(), [listing("r", [{:cnk, "1111111"}, @gtin])], @w1)

      # Same single listing, same barcode — but the national code has been swapped underneath it.
      events = decide(ledger, [listing("r", [{:cnk, "2222222"}, @gtin])], @w2)

      assert Enum.any?(events, &match?(%Events.ConflictFlagged{subject: {:identity_swap, "SK_1"}}, &1))
    end

    test "gaining a second national code is an alias, not a swap, and is not flagged" do
      ledger = week(IdentityLedger.new(), [listing("r", [{:cnk, "1111111"}, @gtin])], @w1)
      events = decide(ledger, [listing("r", [{:cnk, "1111111"}, {:cnk, "2222222"}, @gtin])], @w2)

      refute Enum.any?(events, &match?(%Events.ConflictFlagged{subject: {:identity_swap, _}}, &1))
    end

    test "an ordinary attribute-only change does not flag anything" do
      ledger = week(IdentityLedger.new(), [listing("r", [{:cnk, "1111111"}, @gtin])], @w1)
      events = decide(ledger, [listing("r", [{:cnk, "1111111"}, @gtin])], @w2)

      refute Enum.any?(events, &match?(%Events.ConflictFlagged{}, &1))
    end
  end

  # ── helpers ─────────────────────────────────────────────────────────────────

  defp listing(ref, codes), do: claim(:nhsbsa, :identity, %{ref: ref, codes: codes}, @w1, @w1)

  defp codes(list), do: MapSet.new(Enum.map(list, &Codes.canonicalize/1))

  defp decide(ledger, claims, at) do
    stamped = Enum.map(Enum.with_index(claims, 1), fn {c, i} -> %{c | order: i} end)
    IdentityLedger.decide(ledger, {:reconcile, Cluster.variants(Substrate.current(stamped)), at})
  end

  defp week(ledger, claims, at) do
    ledger
    |> decide(claims, at)
    |> Enum.reduce(ledger, &IdentityLedger.evolve(&2, &1))
  end
end
